Aspect ratio: update to this decade (#16209)
mdn/content: files/en-us/web/media/images/aspect_ratio_mapping/index.md

  - jank
  - responsive
  - width
  - performance
---

> When the `width` and `height` attributes of an image are included on an HTML {{htmlelement("img")}} element, the  aspect ratio of the image can be calculated by the browser prior to the image being loaded. This aspect ratio is used to correct reserve the space needed to display the image preventing a page repaint, which is good for performance. 

Reviewer: ```suggestion
When the `width` and `height` attributes of an image are included on an HTML {{htmlelement("img")}} element, the aspect ratio of the image can be calculated by the browser prior to the image being loaded. This aspect ratio is used to correct and reserve the space needed to display the image preventing a page repaint, which is good for performance. 
```

> When the `width` and `height` attributes of an image are included on an HTML {{htmlelement("img")}} element, the aspect ratio of the image can be calculated by the browser prior to the image being loaded. This aspect ratio is used to reserve the space needed to display the image, reducing or even preventing a layout shift when the image is downloaded and painted to the screen. Reducing layout shift is major component of good user experience and web performance. 

Reviewer: ```suggestion
When the `width` and `height` attributes of an image are included on an HTML {{htmlelement("img")}} element, the aspect ratio of the image can be calculated by the browser prior to the image being loaded. This aspect ratio is used to reserve the space needed to display the image, reducing or even preventing a layout shift when the image is downloaded and painted to the screen. Reducing layout shift is a major component of good user experience and web performance. 
```

> When the `width` and `height` attributes of an image are included on an HTML {{htmlelement("img")}} element, the  aspect ratio of the image can be calculated by the browser prior to the image being loaded. This aspect ratio is used to reserve the space needed to display the image, reducing or even preventing a layout shift when the image is downloaded and painted to the screen. Reducing layout shigt is major component of good user experience and web performance. 

Reviewer: ```suggestion
When the `width` and `height` attributes of an image are included on an HTML {{htmlelement("img")}} element, the aspect ratio of the image can be calculated by the browser prior to the image being loaded. This aspect ratio is used to reserve the space needed to display the image, reducing or even preventing a layout shift when the image is downloaded and painted to the screen. Reducing layout shift is major component of good user experience and web performance. 
```

## Jank problems when loading images

> It is good practice to add `width` and `height` attributes to your HTML {{htmlelement("img")}} elements, so that when browsers first loaded the page, they can reserve a correctly-sized placeholder box in the layout for each image to appear in when it finally loads.

Reviewer: ```suggestion
It is a good practice to add `width` and `height` attributes to your HTML {{htmlelement("img")}} elements, so that when browsers first load the page, they can reserve a correctly-sized placeholder box in the layout for each image to appear in when it finally loads.
```

Author: went with "render" instead of load.

> Browsers beging rendering content as HTML is parsed, often before all assets, including images, are downloaded. It is a good practice to add `width` and `height` attributes to your HTML {{htmlelement("img")}} elements, so that when browsers first render the page, they can reserve a correctly-sized placeholder boxes for each image to appear in when it loaded.

Reviewer: ```suggestion
Browsers begin rendering content as HTML is parsed, often before all assets, including images, are downloaded. It is a good practice to add `width` and `height` attributes to your HTML {{htmlelement("img")}} elements, so that when browsers first render the page, they can reserve a correctly-sized placeholder box for each image to appear in when the images are loaded.
```
there is one box per image, right?

![Two screenshots the first without an image but with space reserved, the second showing the image loaded into the reserved space.](ar-guide.jpg)

> Without the `width` and `height` attributes, no placeholder space is created, creating a noticable {{glossary('jank')}} in the page layout when the image is finally loaded. Page reflow and repaints are performance and usability issues.   

Reviewer: ```suggestion
Without the `width` and `height` attributes, no placeholder space is created, creating a noticeable {{glossary('jank')}} in the page layout when the image is finally loaded. Page reflow and repaints are performance and usability issues.
```

> Without the `width` and `height` attributes, no placeholder space is created, creating a noticable {{glossary('jank')}}, or layout shift, in the page when the image loads after the page is rendered. Page reflow and repaints are performance and usability issues.   

Reviewer: ```suggestion
Without the `width` and `height` attributes, no placeholder space is created, creating a noticeable {{glossary('jank')}}, or layout shift, in the page when the image loads after the page is rendered. Page reflow and repaints are performance and usability issues.
```

> To keep images from breaking out of their containers in responsive designs when the container becomes narrower than the image, the following CSS can be used:

Reviewer: ```suggestion
In responsive designs, when the container becomes narrower than the image, to keep images from breaking out of their containers , the following CSS can be used:
```

Author: manually added because i messed up git.

```css
img, input[type="image"], video, embed, iframe, marquee, object, table {
  aspect-ratio: attr(width) / attr(height);
}
```

As the selector demonstrates, this user-agent style affects any element acting as a container for complex or mixed visual media — {{htmlelement("embed")}}, {{htmlelement("iframe")}}, {{htmlelement("marquee")}}, {{htmlelement("object")}}, {{htmlelement("table")}}, and {{htmlelement("video")}}, in addition to actual images ({{htmlelement("img")}} and `<input type="image">`). When such an element has `width` and `height` attributes set on it, its aspect ratio is calculated before load time, and is available to the browser, using the dimensions provided.

The aspect ratio is then used to calculate the height and therefore the correct size is applied to the `<img>` element, meaning that the aforementioned jank will not occur, or be minimal if the listed dimensions are not fully accurate, when the image loads.

> The aspect ratio is only used to reserve space for the image on load. Once the image has loaded, the intrinsic aspect ratio of the loaded image, rather than the aspect ratio from the attributes, is used. This ensures that it displays at the correct aspect ratio even if the attribute dimensions are not accurate.

Reviewer: ```suggestion
The aspect ratio is used to reserve space only on the image load. Once the image has loaded, the intrinsic aspect ratio of the loaded image, rather than the aspect ratio from the attributes, is used. This ensures that it displays at the correct aspect ratio even if the attribute dimensions are not accurate.
```

}
```

The aspect ratio is then used to calculate the height and therefore the correct size is applied to the `<img>` element, meaning that the aforementioned jank will not occur, or be minimal if the listed dimensions are not fully accurate, when the image loads.

> The aspect ratio is only used to reserve space for the image on load. Once the image has loaded, the intrinsic aspect ratio of the loaded image, rather than the aspect ratio from the attributes is uses, ensuring it displays at the correct aspect ratio.

Reviewer: ```suggestion
The aspect ratio is used to reserve space only on the image load. Once the image has loaded, the intrinsic aspect ratio of the loaded image, rather than the aspect ratio from the attributes is uses, ensuring it displays at the correct aspect ratio.
```

The aspect ratio is then used to calculate the height and therefore the correct size is applied to the `<img>` element, meaning that the aforementioned jank will not occur when the image loads.

The aspect ratio is only used to reserve space for the image on load. Once the image has loaded, the intrinsic aspect ratio of the loaded image rather than the aspect ratio from the attributes is use, ensuring it displays at the correct aspect ratio.

> While CSS best practices from the last decade may have suggested omitting the `width` and `height` attributes of an image on an HTML {{htmlelement("img")}}, due to aspect ration mapping, including these two attributes is once again a best practic.

Reviewer: ```suggestion
While CSS best practices from the last decade may have suggested omitting the `width` and `height` attributes of an image on an HTML {{htmlelement("img")}}, due to aspect ration mapping, including these two attributes is once again a best practice.
```
